tcljava/doc/TclObject.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: TclObject.3,v 1.1 1998/10/14 21:09:09 cvsadmin Exp $
'\" 
.so man.macros
.TH TclObject 3 1.0 Tcl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclObject -- the basic notion of an "object" in Tcl.

.SH CONSTRUCTORS
TclObject

.SH METHODS
.sp
\fBTclObject\fR(InternalRep \fIrep\fR)
.sp
void \fBsetInternalRep\fR(InternalRep \fIrep\fR)
.sp
InternalRep \fBgetInternalRep\fR()
.sp
String \fBtoString\fR()
.sp
void \fBinvalidateStringRep\fR()
.sp
void \fBpreserve\fR()
.sp
void \fBrelease\fR()
.sp
boolean \fBisShared\fR()
.sp
TclObject \fBtakeExclusive\fR()

.SH ARGUMENTS
.AS InternalRep rep .
.AP InternalRep rep
Internal representation to be stored in newly created TclObject.
.BE

.SH DESCRIPTION
The Java API to the Tcl interpreter is object-based -- in most cases,
the values passed to and from the Tcl interpreter are instances of
TclObject. This includes variable values and command arguments.
.PP
A TclObject is dual-ported: it behaves like a String but also holds an
internal representation that can be manipulated more efficiently.  For
example, a TclList is represented as a TclObject that holds the
list's string representation as well as a Vector to hold the
objects for each list element. Dual-ported objects avoid most runtime
type conversions.  This improves the speed of many operations
since an appropriate representation is immediately available.
.PP
Currently, the following internal representations are supported:
TclBoolean, TclDouble, TclList, TclIndex, TclInteger, TclString and
ReflectObject. Most of these internal representations have a
newInstance() method, which can be used to create a new TclObject
instance that contains the specific internal representation.  You
should always create TclObject instances with the newInstance() 
methods; use the "new" operator to create TclObject instances only
when you are writing new internal representations.
.PP
The type of the internal representation of Tcl objects can mutate.
Methods such as \fBsetInternalRep\fR and \fBinvalidateStringlRep\fR
are facilities for mutating the internal rep.
.PP
To improve memory efficiency, TclObject supports copy-on-write
operations.  When you need to save the value of a TclObject for later
use, call the \fBpreserve\fR method; when you no longer need its
value, call the \fBrelease\fR method. Internally, each call to
\fBpreserve\fR will internally increment the "reference count" of the
TclObject by one; conversely, each call to \fBrelease\fR decrements
the reference count by one.
.PP
Some methods of classes derrived from the \fBInternalRep\fR class,
such as \fBTclList\fR.\fBappend()\fR and \fBTclInteger\fR.\fBset()\fR,
modify the contents of a TclObject.  When you are about to call these
methods to modify the value of a TclObject, you must call \fBpreserve\fR (if
you haven't already done so) and then call \fBtakeExclusive\fR.
\fBtakeExclusive\fR will make a copy of the TclObject if it is shared.
The TclObject returned by the \fBtakeExclusive\fR will always have a
reference count of 1.
.TP
\fBTclObject\fR
The \fBTclObject\fR constructor creates a new TclObject with the
internal representation of \fIrep\fR.
.TP
\fBsetInternalRep\fR
The \fBsetInternalRep\fR method changes the internal representation
of the TclObject to the value \fIrep\fR.
.TP
\fBgetInternalRep\fR
The \fBgetInternalRep\fR method returns the internal representation
of the TclObject.  This methods allows you to query the class of a
returned object to determine what type of TclObject it is.
.TP
\fBtoString\fR
The \fBtoString\fR method returns the string representation of the
TclObject.
.TP
\fBinvalidateStringRep\fR
The \fBinvalidateStringRep\fR method marks the String representation
of the TclObject as invalid.  This method should be called only by
subclasses of the InternalRep prior to a call the
\fBsetInternalRep\fR.
.TP
\fBpreserve\fR
The \fBpreserve\fR method saves the value of a TclObject for
later use.  Each call to \fBpreserve\fR will internally increment the
"reference count" of the TclObject by one.
.TP
\fBrelease\fR
After saving the value of a TclObject for later use (via the
\fBpreserve\fR method), call the \fBrelease\fR method when you no
longer need the TclObject's value.  Each call to \fBrelease\fR will
internally decrement the "reference count" of the TclObject by one.
.TP
\fBisShared\fR
The \fBisShared\fR method returns true if the TclObject is shared (its
reference count is greater than one.)
.TP
\fBtakeExclusive\fR
The \fBtakeExclusive\fR method makes a copy of the TclObject if it is
shared.  The TclObject returned by the \fBtakeExclusive\fR will always
have a reference count of 1.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_NewObj, Tcl_ConvertToType, TclObj.typePtr, Tcl_GetStringFromObj,
Tcl_InvalidateStringRep, Tcl_IncrRefCount,
Tcl_DecrRefCount, Tcl_IsShared

.SH "SEE ALSO"
InternalRep, TclBoolean, TclDouble, TclList, TclIndex, TclInteger,
TclString, ReflectObject, Interp

.SH KEYWORDS
internal representation, object, object creation, object type,
reference counting, string representation, type conversion


tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tcljava/tcl/lang/TclObject.java
/*
 * TclObject.java
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclObject.java,v 1.11 2005/09/30 02:12:17 mdejong Exp $
 *
 */

package tcl.lang;

import java.util.Hashtable;
import java.util.Enumeration;

/**
 * This class implements the basic notion of an "object" in Tcl. The
 * fundamental representation of an object is its string value. However,
 * an object can also have an internal representation, which is a "cached"
 * reprsentation of this object in another form. The type of the internal
 * rep of Tcl objects can mutate. This class provides the storage of the
 * string rep and the internal rep, as well as the facilities for mutating
 * the internal rep.
 */

public final class TclObject {
    // Internal representation of the object.

    protected InternalRep internalRep;

    // Reference count of this object. When 0 the object will be deallocated.

    protected int refCount;

    // String  representation of the object.

    protected String stringRep;

    // Setting to true will enable a feature that keeps
    // track of TclObject allocation, internal rep allocation,
    // and transitions from one internal rep to another.

    static final boolean saveObjRecords = false;
    static Hashtable objRecordMap = ( saveObjRecords ? new Hashtable() : null );

    /**
     * Creates a TclObject with the given InternalRep. This method should be
     * called only by an InternalRep implementation.
     *
     * @param rep the initial InternalRep for this object.
     */
    public TclObject(InternalRep rep) {
	if (rep == null) {
	    throw new TclRuntimeError("null InternalRep");
	}
	internalRep = rep;
	stringRep = null;
	refCount = 0;

	if (TclObject.saveObjRecords) {
	    String key = "TclObject";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = new Integer(1);
	    } else {
	        num = new Integer(num.intValue() + 1);
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    /**
     * Creates a TclObject with the given InternalRep and stringRep.
     * This constructor is used by the TclString class only. No other place
     * should call this constructor.
     *
     * @param rep the initial InternalRep for this object.
     * @param s the initial string rep for this object.
     */
    protected TclObject(TclString rep, String s) {
	if (rep == null) {
	    throw new TclRuntimeError("null InternalRep");
	}
	internalRep = rep;
	stringRep = s;
	refCount = 0;

	if (TclObject.saveObjRecords) {
	    String key = "TclObject";
	    Integer num = (Integer) TclObject.objRecordMap.get(key);
	    if (num == null) {
	        num = new Integer(1);
	    } else {
	        num = new Integer(num.intValue() + 1);
	    }
	    TclObject.objRecordMap.put(key, num);
	}
    }

    /**
     * Returns the handle to the current internal rep. This method should be
     * called only by an InternalRep implementation.
     *
     * @return the handle to the current internal rep.
     */
    public final InternalRep getInternalRep() {
	disposedCheck();
	return internalRep;
    }

    /**
     * Change the internal rep of the object. The old internal rep
     * will be deallocated as a result. This method should be
     * called only by an InternalRep implementation.
     *
     * @param rep the new internal rep.
     */
    public final void setInternalRep(InternalRep rep) {
	disposedCheck();
	if (rep == null) {
	    throw new TclRuntimeError("null InternalRep");
	}
	if (rep == internalRep) {
	    return;
	}

	// In the special case where the internal representation is a CObject,
	// we want to call the special interface to convert the underlying
	// native object into a reference to the Java TclObject.  Note that
	// this test will always fail if we are not using the native
	// implementation. Also note that the makeReference method
	// will do nothing in the case where the Tcl_Obj inside the
	// CObject was originally allocated in Java. When converting
	// to a CObject we need to break the link made earlier.

	if ((internalRep instanceof CObject) && !(rep instanceof CObject)) {
	    // We must ensure that the string rep is copied into Java
	    // before we lose the reference to the underlying CObject.
	    // Otherwise we will lose the original string information
	    // when the backpointer is lost.

	    if (stringRep == null) {
		stringRep = internalRep.toString();
	    }
	    ((CObject) internalRep).makeReference(this);
	}

        //System.out.println("TclObject setInternalRep for \"" + stringRep + "\"");
        //System.out.println("from \"" + internalRep.getClass().getName() +
        //    "\" to \"" + rep.getClass().getName() + "\"");
	internalRep.dispose();
	internalRep = rep;
    }

    /**
     * Returns the string representation of the object.
     *
     * @return the string representation of the object.
     */
    public final String toString() {
	disposedCheck();
	if (stringRep == null) {
	    stringRep = internalRep.toString();
	}
	return stringRep;
    }

    /**
     * Sets the string representation of the object to null.  Next
     * time when toString() is called, getInternalRep().toString() will
     * be called. This method should be called ONLY when an InternalRep
     * is about to modify the value of a TclObject.
     *
     * @exception TclRuntimeError if object is not exclusively owned.
     */
    public final void invalidateStringRep() throws TclRuntimeError {
	disposedCheck();
	if (refCount > 1) {
	    throw new TclRuntimeError("string representation of object \"" +
		    toString() + "\" cannot be invalidated: refCount = " +
		    refCount);
	}
	stringRep = null;
    }

    /**
     * Returns true if the TclObject is shared, false otherwise.
     * @return true if the TclObject is shared, false otherwise.
     */
    public final boolean isShared() {
	disposedCheck();
	return (refCount > 1);
    }

    /**
     * Tcl_DuplicateObj -> duplicate
     *
     * Duplicate a TclObject, this method provides the preferred
     * means to deal with modification of a shared TclObject.
     * It should be invoked in conjunction with isShared instead
     * of using the deprecated takeExclusive method.
     *
     * Example:
     *
     *		if (tobj.isShared()) {
     *		    tobj = tobj.duplicate();
     *		}
     *		TclString.append(tobj, "hello");
     *
     * @return an TclObject with a refCount of 0.
     */

    public final TclObject duplicate() {
	disposedCheck();
	if (internalRep instanceof TclString) {
	    if (stringRep == null) {
	        stringRep = internalRep.toString();
	    }
	}
	TclObject newObj = new TclObject(internalRep.duplicate());
	newObj.stringRep = this.stringRep;
	newObj.refCount = 0;
	return newObj;
    }

    /**
     * @deprecated The takeExclusive method has been deprecated
     * in favor of the new duplicate() method. The takeExclusive
     * method would modify the ref count of the original object
     * and return an object with a ref count of 1 instead of 0.
     * These two behaviors lead to lots of useless duplication
     * of objects that could be modified directly.
     */

    public final TclObject takeExclusive() throws TclRuntimeError {
	disposedCheck();
	if (refCount == 1) {
	    return this;
	} else if (refCount > 1) {
	    if (internalRep instanceof TclString) {
		if (stringRep == null) {
		    stringRep = internalRep.toString();
		}
	    }
	    TclObject newObj = new TclObject(internalRep.duplicate());
	    newObj.stringRep = this.stringRep;
	    newObj.refCount = 1;
	    refCount--;
	    return newObj;
	} else {
	    throw new TclRuntimeError("takeExclusive() called on object \"" +
		    toString() + "\" with: refCount = 0");
	}
    }

    /**
     * Tcl_IncrRefCount -> preserve
     *
     * Increments the refCount to indicate the caller's intent to
     * preserve the value of this object. Each preserve() call must be matched
     * by a corresponding release() call.
     *
     * @exception TclRuntimeError if the object has already been deallocated.
     */
    public final void preserve() throws TclRuntimeError {
	disposedCheck();
	if (internalRep instanceof CObject) {
	    ((CObject) internalRep).incrRefCount();
	}
        _preserve();
    }

    /**
     * _preserve
     *
     * Private implementation of preserve() method.
     * This method will be invoked from Native code
     * to change the TclObject's ref count without
     * effecting the ref count of a CObject.
     */
    private final void _preserve() throws TclRuntimeError {
	refCount++;
    }

    /**
     * Tcl_DecrRefCount -> release
     *
     * Decrements the refCount to indicate that the caller is no longer
     * interested in the value of this object. If the refCount reaches 0,
     * the obejct will be deallocated.
     */
    public final void release() {
	disposedCheck();
	if (internalRep instanceof CObject) {
	    ((CObject) internalRep).decrRefCount();
	}
	_release();
    }

    /**
     * _release
     *
     * Private implementation of preserve() method.
     * This method will be invoked from Native code
     * to change the TclObject's ref count without
     * effecting the ref count of a CObject.
     */
    private final void _release() {
	refCount--;
	if (refCount <= 0) {
	    internalRep.dispose();

	    // Setting these to null will ensure that any attempt to use
	    // this object will result in a Java NullPointerException.

	    internalRep = null;
	    stringRep = null;
	}
    }

    /**
     * Returns the refCount of this object.
     *
     * @return refCount.
     */
    final int getRefCount() {
	return refCount;
    }

    /**
     * Returns the Tcl_Obj* objPtr member for a CObject or TclList.
     * This method is only called from Tcl Blend.
     */

    final long getCObjectPtr() {
	if (internalRep instanceof CObject) {
	    return ((CObject) internalRep).getCObjectPtr();
	} else {
	    return 0;
	}
    }

    /**
     * Returns 2 if the internal rep is a TclList.
     * Returns 1 if the internal rep is a CObject.
     * Otherwise returns 0.
     * This method provides an optimization over
     * invoking getInternalRep() and two instanceof
     * checks via JNI. It is only used by Tcl Blend.
     */

    final int getCObjectInst() {
	if (internalRep instanceof CObject) {
	    if (internalRep instanceof TclList)
	        return 2;
	    else
	        return 1;
	} else {
	    return 0;
	}
    }

    /**
     * Raise a TclRuntimeError if this TclObject has been
     * disposed of because the last ref was released.
     */

    private final void disposedCheck() {
	if (internalRep == null) {
	    throw new TclRuntimeError("TclObject has been deallocated");
	}
    }

    /**
     * Return a String that describes TclObject and internal
     * rep type allocations and conversions. The string is
     * in lines seperated by newlines. The saveObjRecords
     * needs to be set to true and Jacl recompiled for
     * this method to return a useful value.
     */

    public static String getObjRecords() {
        if (TclObject.saveObjRecords) {
            StringBuffer sb = new StringBuffer(64);
            for ( Enumeration keys = TclObject.objRecordMap.keys() ;
                    keys.hasMoreElements() ; ) {
                String key = (String) keys.nextElement();
                Integer num = (Integer) TclObject.objRecordMap.get(key);
                sb.append(key);
                sb.append(" ");
                sb.append(num.intValue());
                sb.append("\n");
            }
            TclObject.objRecordMap = new Hashtable();
            return sb.toString();
        } else {
            return "";
        }
    }

    // Return true if this TclObject has no string rep.
    // This could be the case with a "pure" integer
    // or double, or boolean, that was created from
    // a primitive value. Once the toString() method
    // has been invoked, this method will return true
    // unless the string rep is again invalidated.

    final boolean hasNoStringRep() {
        return (stringRep == null);
    }
}

